### PR TITLE
Improve the determination of cluster exemplars fix for main

### LINKED ISSUE #355: HDBSCAN implementation in 4.3+
Hey,

Thanks for your work on this library and for making this open-source.

I was testing different implementations of HDBSCAN, and made a trivial test like this for Tribuo:

```
public void test() {
       final HdbscanTrainer trainer = new HdbscanTrainer(
                100,
                HdbscanTrainer.Distance.EUCLIDEAN,
                10,
                1
        );

        var outputFactory = new ClusteringFactory();

        var dataProvenance = new SimpleDataSourceProvenance(
                "test",
                outputFactory
        );

        var dataList = new ArrayList<Example<ClusterID>>();

        for (var n = 0; n < 10; n++) {
            for (double i = 1.0; i <= 2.0; i += 0.1) {
                dataList.add(createExample(i));
            }
            for (double i = 10.0; i <= 20.0; i += 0.1) {
                dataList.add(createExample(i));
            }
        }

        var dataSource = new ListDataSource(
                dataList,
                outputFactory,
                dataProvenance
        );

        var trainData = new MutableDataset<ClusterID>(
                dataSource
        );
        
        HdbscanModel model = trainer.train(trainData);
        System.out.printf("Num clusters: %d%n", new HashSet<>(model.getClusterLabels()).size());
        
        // Predicting on the same training data
        var predictions = model.predict(trainData);

        for (var prediction : predictions) {
            System.out.printf("%s: %d %f\n",prediction.getExample(), prediction.getOutput().getID(), prediction.getOutput().getScore());
        }
}
        
private Example<ClusterID> createExample(double value) {
        var example = new ArrayExample<ClusterID>(
                ClusteringFactory.UNASSIGNED_CLUSTER_ID,
                new String[]{"A"},
                new double[]{value}
        );
        return example;
}
```

Now surprisingly, version 4.2 worked as expected(?) identifying 2 clusters and prediction work as well.
For version 4.3 exactly the same code above identifies also 2 clusters, but the prediction returns all as outliers (ClusterID: 0  with score: 0.999900).

Am I missing something or HDBSCAN implementation in 4.3 gives unexpected results?

**Is your question about a specific ML algorithm or approach?**
HDBSCAN implementation in Tribuo.

**System details**
- Tribuo version: 4.2.2, 4.3.1

### PR DESCRIPTION
**Pull request by geoffreydstewart**
### Description
We need to improve the determination of cluster exemplars, which are used after the model is trained to make predictions. Some cases, such as those which use contrived datasets, can result in very "clean" clusters, with no outliers. One such case has exposed an issue in the current logic to determine cluster exemplars. The seed, which is used when sampling cluster exemplars at random from members of a cluster, is exposed as a config parameter.

### Motivation
This change fixes #355 on the main branch.

### CODE REVIEW

File: Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java [hunk 1]
```diff
@@ -779,31 +804,68 @@ private static List<ClusterExemplar> computeExemplars(SGDVector[] data, Map<Inte
...
+                    Stream<Integer> intStreamTree = IntStream.range(0, outlierScoreTreeSize).boxed();
+                    intStreamTree.forEach((i) -> {
+                        Entry<Double, Integer> treeEntry = outlierScoreIndexTree.pollFirstEntry();
+                        partialClusterExemplarsList.add(new Pair<>(treeEntry.getKey(), treeEntry.getValue()));
+                    });
+
+                    // To determine the remaining exemplars, the best thing to do is randomly sample them from all the
+                    // points in this cluster. This could introduce duplicate exemplar points, but that is safer than
+                    // reducing the number of exemplars.
+                    SplittableRandom rand = new SplittableRandom(exemplarSampleSeed);
```
Comment (Craigacp) [inline Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java:849]: How much of an edge case is this randomized fallback? We have special treatment for all other RNGs in Tribuo where they are members of the trainer and split under a lock to preserve provenance information, which ensures that repeated runs of a trainer on the same data give different answers when not controlling the RNG state (e.g. when used in an ensemble). If this is purely an edge case issue then it might be ok to leave it as is, but if it might occur relatively frequently then it's probably worth using the same idiom we do elsewhere.

Comment (geoffreydstewart) [reply Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java:849]: This edge case seems to be quite rare, but if there is a "best solution" which can be implemented here we should strive towards that. I'll assume that the implementation of this special treatment demonstrated in the `KMeansTrainer` is a good example to follow.

Comment (Craigacp) [reply Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java:849]: Yeah, this synchronized idiom along with storing an rng as a field - https://github.com/oracle/tribuo/blob/main/Clustering/KMeans/src/main/java/org/tribuo/clustering/kmeans/KMeansTrainer.java#L278.

Comment (geoffreydstewart) [reply Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java]: I've added a commit which adds a RNG as a field, and the supporting logic. Let me know if anything else might be needed.

File: Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java [hunk 2]
```diff
@@ -150,6 +151,12 @@ public DistanceType getDistanceType() {
     @Config(description = "The nearest neighbour implementation factory to use.")
     private NeighboursQueryFactory neighboursQueryFactory;
 
+    @Config(description = "The seed to use in the event cluster exemplars need to be determined with random samples " +
+        "from the members of a cluster.")
+    private long exemplarSampleSeed = Trainer.DEFAULT_SEED;
+
+    private SplittableRandom rng = new SplittableRandom(exemplarSampleSeed);
```
[the remark this answers is not in the document]
Comment (geoffreydstewart) [reply Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java:158]: Right, I've made this change as well as adding the call to `postConfig` from the Constructors.

**Comment by geoffreydstewart** [comment]
> Needs a fix to the RNG creation, otherwise looks good.

Just following up to see what else might be needed here, no rush though.

**Review by Craigacp — APPROVED** [review]
LGTM, thanks.
